From 892d306ba0304a31a0b272585b52488c9ee69462 Mon Sep 17 00:00:00 2001
From: Ahmad Samir <a.samirh78@gmail.com>
Date: Mon, 13 Oct 2025 08:22:11 +0300
Subject: [PATCH 449/553] tst_QSettings: split the testEscapes() test:
 testEscapedStringList

Easier to debug.

Use example.org instead of software.org, the latter is regitered and we
shouldn't use it in tests.

Pick-to: 6.8 6.5
Change-Id: I12df40ed1925a95976a75011f24ed95facb98a19
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
(cherry picked from commit 67eb96fe8d112a0d2ab322ebc7f50dd06a08d2f9)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../corelib/io/qsettings/tst_qsettings.cpp    | 77 +++++++++++++------
 1 file changed, 54 insertions(+), 23 deletions(-)

diff --git a/tests/auto/corelib/io/qsettings/tst_qsettings.cpp b/tests/auto/corelib/io/qsettings/tst_qsettings.cpp
index c981dc366b0..46a89c5631c 100644
--- a/tests/auto/corelib/io/qsettings/tst_qsettings.cpp
+++ b/tests/auto/corelib/io/qsettings/tst_qsettings.cpp
@@ -188,6 +188,8 @@ private slots:
     void testEscapedKeys();
     void testUnescapedKeys_data();
     void testUnescapedKeys();
+    void testEscapedStringList_data();
+    void testEscapedStringList();
     void testNormalizedKey_data();
     void testNormalizedKey();
     void testVariantTypes_data() { populateWithFormats(); }
@@ -2834,15 +2836,6 @@ void tst_QSettings::testEscapes()
 {
     QSettings settings(QSettings::UserScope, "software.org", "KillerAPP");
 
-#define testEscapedStringList(plainStrList, escStrList) \
-    { \
-        QStringList plainList(plainStrList); \
-        QByteArray escList(escStrList); \
-        QCOMPARE(iniEscapedStringList(plainList), escList); \
-        QCOMPARE(iniUnescapedStringList(escList), plainList); \
-    } \
-
-
 #define testUnescapedStringList(escStrList, plainStrList, reescStrList) \
     { \
         QStringList plainList(plainStrList); \
@@ -2869,20 +2862,6 @@ void tst_QSettings::testEscapes()
         QCOMPARE(v.toString(), QString(vStr)); \
     }
 
-    testEscapedStringList("", "");
-    testEscapedStringList(" ", "\" \"");
-    testEscapedStringList(";", "\";\"");
-    testEscapedStringList(",", "\",\"");
-    testEscapedStringList("=", "\"=\"");
-    testEscapedStringList("abc-def", "abc-def");
-    testEscapedStringList(QChar(0) + QString("0"), "\\0\\x30");
-    testEscapedStringList("~!@#$%^&*()_+.-/\\=", "\"~!@#$%^&*()_+.-/\\\\=\"");
-    testEscapedStringList("~!@#$%^&*()_+.-/\\", "~!@#$%^&*()_+.-/\\\\");
-    testEscapedStringList(QString("\x7F") + "12aFz", QByteArray("\x7f") + "12aFz");
-    testEscapedStringList(QString("   \t\n\\n") + QChar(0x123) + QChar(0x4567), "\"   \\t\\n\\\\n\xC4\xA3\xE4\x95\xA7\"");
-    testEscapedStringList(QString("\a\b\f\n\r\t\v'\"?\001\002\x03\x04"), "\\a\\b\\f\\n\\r\\t\\v'\\\"?\\x1\\x2\\x3\\x4");
-    testEscapedStringList(QStringList() << "," << ";" << "a" << "ab,  \tc, d ", "\",\", \";\", a, \"ab,  \\tc, d \"");
-
     /*
       Test .ini syntax that cannot be generated by QSettings (but can be entered by users).
     */
@@ -2992,6 +2971,58 @@ void tst_QSettings::testUnescapedKeys()
     QCOMPARE(iniUnescapedKey(reescKey), plainKey);
 }
 
+void tst_QSettings::testEscapedStringList_data()
+{
+    QTest::addColumn<QStringList>("plainStrList");
+    QTest::addColumn<QByteArray>("escapedList");
+
+    QTest::newRow("empty-string") << QStringList{u""_s} << ""_ba;
+    QTest::newRow("space") << QStringList{u" "_s} << "\" \""_ba;
+    QTest::newRow(";") << QStringList{u";"_s} << "\";\""_ba;
+    QTest::newRow(",") << QStringList{u","_s} << "\",\""_ba;
+    QTest::newRow("=") << QStringList{u"="_s} << "\"=\""_ba;
+    QTest::newRow("abc-def") << QStringList{u"abc-def"_s} << "abc-def"_ba;
+
+    QTest::newRow("starts-with-NUL")
+        << QStringList{QChar(0) + u"0"_s}
+        << "\\0\\x30"_ba;
+
+    QTest::newRow("special-characters1")
+        << QStringList{u"~!@#$%^&*()_+.-/\\="_s}
+        << "\"~!@#$%^&*()_+.-/\\\\=\""_ba;
+
+    QTest::newRow("special-characters2")
+        << QStringList{u"~!@#$%^&*()_+.-/\\"_s}
+        << "~!@#$%^&*()_+.-/\\\\"_ba;
+
+    QTest::newRow("DEL-character")
+        << QStringList{u"\x7F"_s + u"12aFz"_s}
+        << "\x7f"_ba + "12aFz"_ba;
+
+    QTest::newRow("tab-newline")
+        << QStringList{u"   \t\n\\n"_s + QChar(0x123) + QChar(0x4567)}
+        << "\"   \\t\\n\\\\n\xC4\xA3\xE4\x95\xA7\""_ba;
+
+    QTest::newRow("backslash-espcaped-input")
+        << QStringList{u"\a\b\f\n\r\t\v'\"?\001\002\x03\x04"_s}
+        << "\\a\\b\\f\\n\\r\\t\\v'\\\"?\\x1\\x2\\x3\\x4"_ba;
+
+    QTest::newRow("stringlist-with-tab")
+        << QStringList{u","_s, u";"_s, u"a"_s, u"ab,  \tc, d "_s}
+        << "\",\", \";\", a, \"ab,  \\tc, d \""_ba;
+}
+
+void tst_QSettings::testEscapedStringList()
+{
+    QFETCH(QStringList, plainStrList);
+    QFETCH(QByteArray, escapedList);
+
+    QSettings settings(QSettings::UserScope, "example.org", "KillerAPP");
+
+    QCOMPARE(iniEscapedStringList(plainStrList), escapedList);
+    QCOMPARE(iniUnescapedStringList(escapedList), plainStrList);
+}
+
 #endif
 
 void tst_QSettings::testCaseSensitivity()
-- 
2.51.2

